src/backend/distributed/utils/node_metadata.c

```diff
@@ -1088,38 +1154,62 @@ AddNodeMetadata(char *nodeName, int32 nodePort,
 		SendCommandToWorkers(WORKERS_WITH_METADATA, nodeInsertCommand);
 	}
 
-	return nextNodeIdInt;
+	return workerNode->nodeId;
 }
 
 
 /*
- * SetNodeState function sets the isactive column of the specified worker in
- * pg_dist_node to isActive.
+ * SetWorkerColumn function sets the column with the specified index
+ * (see pg_dist_node.h) on the worker in pg_dist_node.
+ * It returns the new worker node after the modification.
  */
-static void
-SetNodeState(char *nodeName, int32 nodePort, bool isActive)
+static WorkerNode *
+SetWorkerColumn(WorkerNode *workerNode, int columnIndex, Datum value)
 {
 	Relation pgDistNode = heap_open(DistNodeRelationId(), RowExclusiveLock);
 	TupleDesc tupleDescriptor = RelationGetDescr(pgDistNode);
-	HeapTuple heapTuple = GetNodeTuple(nodeName, nodePort);
+	HeapTuple heapTuple = GetNodeTuple(workerNode->workerName, workerNode->workerPort);
+	WorkerNode *newWorkerNode = NULL;
 
 	Datum values[Natts_pg_dist_node];
 	bool isnull[Natts_pg_dist_node];
 	bool replace[Natts_pg_dist_node];
+	char *metadataSyncCommand = NULL;
 
-	char *nodeStateUpdateCommand = NULL;
-	WorkerNode *workerNode = NULL;
+
+	switch (columnIndex)
+	{
+		case Anum_pg_dist_node_isactive:
+		{
+			metadataSyncCommand = ShouldHaveShardsUpdateCommand(workerNode->nodeId,
+																DatumGetBool(value));
+			break;
+		}
+
+		case Anum_pg_dist_node_shouldhaveshards:
+		{
+			metadataSyncCommand = ShouldHaveShardsUpdateCommand(workerNode->nodeId,
+																DatumGetBool(value));
+			break;
+		}
+
+		default:
+		{
+			ereport(ERROR, (errmsg("could not find valid entry for node \"%s:%d\"",
```

Comment:
This error seems strange for the invalid column index. I think this should be probably replaced with an `Assert(false)`, since this case seems a programming error and should be unreachable.